cdc/puller/buffer.go

```diff
@@ -37,4 +55,112 @@ func (b Buffer) Get(ctx context.Context) (model.RegionFeedEvent, error) {
 	}
 }
 
-// TODO limit memory buffer
+var _ EventBuffer = &memBuffer{}
+
+type memBuffer struct {
+	limitter *BlurResourceLimitter
+
+	mu struct {
+		sync.Mutex
+		entries deque.Deque
+	}
+	signalCh chan struct{}
+}
+
+// Passing nil will make a unlimited  buffer.
+func makeMemBuffer(limitter *BlurResourceLimitter) *memBuffer {
+	return &memBuffer{
+		limitter: limitter,
+		mu: struct {
+			sync.Mutex
+			entries deque.Deque
+		}{
+			entries: deque.NewDeque(),
+		},
+
+		signalCh: make(chan struct{}, 1),
+	}
+}
+
+// AddEntry implements EventBuffer interface.
+func (b *memBuffer) AddEntry(ctx context.Context, entry model.RegionFeedEvent) error {
+	b.mu.Lock()
+	if b.limitter != nil && b.limitter.OverBucget() {
+		b.mu.Unlock()
+		return ErrReachLimit
+	}
+
+	b.mu.entries.PushBack(entry)
```

Comment:
why we use `deque` but not `channel` here?

Reply:
`channel` is a bounded buffer